src/sync/PostSyncProcessor.js

```diff
@@ -0,0 +1,157 @@
+    this.postProcessRequisition = this.postProcessRequisition.bind(this);
+    this.postProcessTransaction = this.postProcessTransaction.bind(this);
+
+    this.database.addListener(this.onDatabaseEvent);
+  }
+
+  /**
+ * Respond to a database change event. Must be called from within a database
+ * write transaction.
+ * @param  {string} changeType  The type of database change, e.g. CREATE, UPDATE, DELETE
+ * @param  {string} recordType  The type of record changed (from database schema)
+ * @param  {object} record      The record changed
+ * @param  {string} causedBy    The cause of this database event, either 'sync' or undefined
+ * @return {none}
+ * Runs all the post sync functions triggered by sync events on database through
+ * this.onDatabaseEvent
+ */
+  onDatabaseEvent(changeType, recordType, record, causedBy) {
+    // Exit if not a change caused by incoming sync
+    if (causedBy !== 'sync' || recordType === 'SyncOut') return;
+    if (this.recordQueue.has(record.id)) {
+      // Check if already in queue, remove old
+      this.recordQueue.delete(record.id);
+    }
+    // Add new entry at end of Map
+    this.recordQueue.set(record.id, recordType);
+  }
+
+  /**
+   * Runs post checks across tables, ensuring data is correct.
+   * Tables manually added as to not iterate over tables that don't have any post processing.
+   */
+  checkTables() {
+    this.actionQueue = [];
+    this.recordQueue = []; // Reset the recordQueue to avoid unnessary runs
+
+    this.database
+      .objects('Requisition')
+      .forEach(record => this.delegateByRecordType('Requisition', record));
+
+    this.database
+      .objects('Transaction')
+      .forEach(record => this.delegateByRecordType('Transaction', record));
+
+    this.runActionQueue();
+  }
+
+  /**
+   * Iterates through records added through listening to sync, adding needed actions
+   * to actionQueue. Runs the actionQueue, making the changes.
+   */
+  processRecordQueue() {
+    this.recordQueue.forEach((recordType, recordId) => {
+      // Use local database record, not what comes in sync. Ensures that records are
+      // integrated information (definitely after sync is done)
+      const internalRecord = this.database.objects(recordType).filtered('id == $0', recordId)[0];
+      this.delegateByRecordType(recordType, internalRecord);
+    });
+    this.runActionQueue();
+    this.recordQueue = []; // Reset the recordQueue to avoid unnessary reruns
+  }
+
+  /**
+   * Runs all the post sync functions triggered by sync events on database through
+   * this.onDatabaseEvent
+   */
+  runActionQueue() {
+    this.database.write(() => this.actionQueue.forEach(func => func()));
+    this.actionQueue = [];
+  }
+
+  /**
+   * Delegates records to the correct utility function
+   * @param  {string} recordType  The type of record changed (from database schema)
+   * @param  {object} record      The record changed
+   * @return {none}
+   */
+  delegateByRecordType(recordType, record) {
```

Comment:
Umm. This looks workable, but not very extendable:
Can this class be made in a way where one can add to a queue of postProcess functions, i.e.:
```JS
this.postSyncProcessor = new PostSyncProcessor(this.database);
this.postSyncProcessor.addProcess('Requisition', (record, database) => {
  if (record.serialNumber === '-1') {
    record.serialNumber =  getNextNumber(database, REQUISITION_SERIAL_NUMBER)
    return true; // to tell postSyncProcessor to do this.database.update(recordType, record));
 } 
  return false;
});
```
In PostSyncProcessor class
```JS
addProcess(recordType, callback) {
  if(!this.processList[recordType]) this.processList[recordType] = [];
  this.processList[recordType].push(callback);
}

onDatabaseEvent(changeType, recordType, record, causedBy) {
...
if (causedBy !== 'sync' || recordType === 'SyncOut' || !Object.keys(this.processList).includes(recordType)) return;
...

  processRecordQueue() {
   this.database.write(() => {
    this.recordQueue.forEach((recordType, recordId) => { // ahh this is confusing for Map(), usually it's (key, value), here it's (value, key).
     
      const internalRecord = this.database.objects(recordType).filtered('id == $0', recordId)[0];
      
      this.dealWithRecordType(recordType, internalRecord);
      this.database.update('Requisition', record));
    });
  });
    this.recordQueue = []; 
  }

dealWithRecordType(recordType, record) {
  if (this.processList[recordType]) {
   this.processList[recordType].forEach(callback => {
   if (callback(record, this.database))
    this.database.update(recordType, record));
   });
  }   
}
```
Of course, code can look tidier then my markdown.

And similar logic for checkTables.

This way PostSyncProcessor becomes dynamic. It's easier to add multiple postSyncProcesses without touching PostSyncProcess logic.